torch/testing/_internal/common_distributed.py

```diff
@@ -465,16 +466,18 @@ class Event(Enum):
         GET_TRACEBACK = 1
 
     @staticmethod
-    def _event_listener(pipe, rank: int):
+    def _event_listener(parent_pipe, signal_pipe, rank: int):
         logger.info(f'Starting event listener thread for {rank}')
         while True:
-            if pipe.poll(None):
+            ready_pipes = multiprocessing.connection.wait([parent_pipe, signal_pipe])
```

Comment:
We probably need a timeout here since this is a non-daemon thread now? We don't want to block here forever?

Reply:
Could this also throw an exception how `pipe.poll` did earlier?

Reply:
> We probably need a timeout here since this is a non-daemon thread now? We don't want to block here forever?

I'm not sure we need a timeout:
- In the "normal" case (no deadlock), the main thread will call `signal_pipe.send(None)` at shutdown, which will make `signal_pipe` readable, causing this poll to return and the thread then to terminate.
- In the "deadlock" case the parent process will anyways end up killing the subprocess ungracefully, and I don't think it matters whether the thread is a daemon or not.

> Could this also throw an exception how pipe.poll did earlier?

It might, though this is not part of the problem. Just like before, those exceptions would clog up the log but be harmless. They also wouldn't interfere with the thread winding down because, well, an uncaught exception kills the thread.